src/test/unit/mcmc/chainset_test.cpp

```diff
@@ -0,0 +1,334 @@
+  EXPECT_EQ(1, chain_1.num_chains());
+
+  // test against R implementation in pkg posterior
+  Eigen::VectorXd rhat_8_schools_1_bulk(10);
+  rhat_8_schools_1_bulk << 1.0012958313, 1.0046136496, 1.0085723580,
+      1.0248629375, 1.0111456620, 1.0004458336, 0.9987162973, 1.0339773469,
+      0.9985612618, 1.0281667351;
+
+  Eigen::VectorXd rhat_8_schools_1_tail(10);
+  rhat_8_schools_1_tail << 1.005676523, 1.009670999, 1.00184184, 1.002222679,
+      1.004148161, 1.003218528, 1.009195353, 1.001426744, 1.003984381,
+      1.025817745;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto rhats = chain_1.split_rank_normalized_rhat(i + 7);
+    EXPECT_NEAR(rhats.first, rhat_8_schools_1_bulk(i), 0.05);
+    EXPECT_NEAR(rhats.second, rhat_8_schools_1_tail(i), 0.05);
+  }
+}
+
+TEST_F(McmcChains, split_rank_normalized_ess) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  // test against R implementation in pkg posterior (via cmdstanr)
+  Eigen::VectorXd ess_8_schools_bulk(10);
+  ess_8_schools_bulk << 348, 370, 600, 638, 765, 608, 629, 274, 517, 112;
+  Eigen::VectorXd ess_8_schools_tail(10);
+  ess_8_schools_tail << 845, 858, 874, 726, 620, 753, 826, 628, 587, 108;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto ess = chain_2.split_rank_normalized_ess(i + 7);
+    EXPECT_NEAR(ess.first, ess_8_schools_bulk(i), 5);
+    EXPECT_NEAR(ess.second, ess_8_schools_tail(i), 5);
+  }
+}
+
+TEST_F(McmcChains, ess_short_chains) {
+  std::vector<stan::io::stan_csv> eight_schools_5iters;
+  eight_schools_5iters.push_back(eight_schools_5iters_1);
+  eight_schools_5iters.push_back(eight_schools_5iters_2);
+  stan::mcmc::chainset chain_2(eight_schools_5iters);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto ess = chain_2.split_rank_normalized_ess(i + 7);
+    EXPECT_TRUE(std::isnan(ess.first));
+    EXPECT_TRUE(std::isnan(ess.second));
+  }
+}
+
+TEST_F(McmcChains, summary_stats) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  // test against R implementation in pkg posterior (via cmdstanr)
+  Eigen::VectorXd s8_mean(10), s8_median(10), s8_sd(10), s8_mad(10), s8_q5(10),
+      s8_q95(10);
+  s8_mean << 7.95, 12.54, 7.82, 5.33, 7.09, 4.12, 5.72, 11.65, 8.80, 8.26;
+  s8_median << 8.00, 11.27, 7.39, 5.44, 6.64, 4.54, 5.93, 11.38, 8.28, 7.05;
+  s8_sd << 5.48, 9.57, 6.85, 8.39, 6.91, 6.57, 6.85, 7.76, 8.40, 5.53;
+  s8_mad << 5.49, 8.79, 6.39, 7.38, 5.98, 6.25, 6.59, 7.79, 7.59, 4.66;
+  s8_q5 << -0.46, -0.39, -3.04, -8.90, -3.31, -7.58, -5.84, 0.10, -4.15, 2.08;
+  s8_q95 << 17.01, 30.47, 19.25, 19.02, 18.72, 14.49, 16.04, 25.77, 22.71,
+      18.74;
```

Comment:
I think this is simply too few decimal places to tell between the various implementations of quantile for this data. 

The one in stan math (added in https://github.com/stan-dev/math/pull/2398) claims to match R's implementation, but past a couple decimal places it gives different answers than this, which also claims to be tested against R's implementation. Math's is still only tested to 3 decimals, so it's hard to say.

If we were going to go with one, I'd say `stan::math::quantile`, but being sure they actually implement the same algorithm would be a nice justification

Reply:
(`stan::math::quantile` is also well defined a 0 and 1., which stansummary currently prevents you from requesting, so we could relax that in this case)